### PR TITLE
Embed data roles in ciphertext nodejs sdk

### PR DESCRIPTION
**Pull request by jakekgrog**
# Why
Data roles need to be embedded in the ciphertext.

# How

- Embed the data role in the ciphertext
- Perform KDF on the new version with Secp256k1 curve
- Add curve parameters for Secp256k1 curve

### CODE REVIEW

File: lib/core/crypto.js [hunk 1]
```diff
@@ -123,13 +132,27 @@ module.exports = (config) => {
...
+        eo: 5, // Encryption origin
+        et: Math.round(Date.now() / 1000), // Encryption timestamp
+      };
+      const metadataBytes = pack(encryptionMetadata);
+      let offsetBuffer = Buffer.allocUnsafe(2);
+      offsetBuffer.writeUInt16LE(metadataBytes.length);
+      result = Buffer.concat([offsetBuffer, metadataBytes, Buffer.from(str)]);
+    } else {
+      result = Buffer.from(str);
+    }
```
Comment (boilsquid) [inline lib/core/crypto.js:152]: Extract to a method, return a Buffer, write a test

Comment (boilsquid) [reply lib/core/crypto.js:152]: This is the way

File: .changeset/many-hairs-bathe.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@evervault/sdk': minor
+---
+
+Add data roles
```
Comment (jeremywagemans) [inline .changeset/many-hairs-bathe.md:5]: We really need to polish our changelogs. I'm suggesting updating this with:
```
The `encrypt` function has been enhanced to accept an optional Data Role. This role, once specified, is associated with the data upon encryption. Data Roles can be created in the Evervault Dashboard (Data Roles section) and provide a mechanism for setting clear rules that dictate how and when data, tagged with that role, can be decrypted.
evervault.encrypt('hello world!', 'allow-all');
```

File: lib/index.js [hunk 3]
```diff
@@ -275,9 +280,10 @@ class EvervaultClient {
 
   /**
    * @param {Object || String} data
+   * @param {Object} metadata
    * @returns {Promise<Object || String>}
    */
-  async encrypt(data) {
+  async encrypt(data, metadata) {
```
Comment (jeremywagemans) [inline lib/index.js:286]: Signature should be `encrypt(data: Any, role: String?)`

Comment (jeremywagemans) [reply lib/index.js:286]: Also could we validate the role? 

```
    const dataRoleRegex = /^[a-z0-9-]{1,20}$/;
    if (role != null && !dataRoleRegex.test(role)) {
      throw new Error('The provided Data Role slug is invalid. The slug can be retrieved in the Evervault dashboard (Data Roles section).');
    }
    ```

**Review by jeremywagemans — CHANGES_REQUESTED** [review]
I added a few suggestions.

A few things are also missing:
- We need to update the end to end tests to validate the encrypt function with data role. I would add encrypt data with the `permit-all` data role (already available in the SDK Tests app).
- We need to ensure the forbidden decryption error is handled correctly. I would add an test case for that in our E2E tests as well (encrypt with `forbid-all` data role).
- Roles are not embedded for files. We need to add E2E test cases for that as well.

**Review by jeremywagemans — APPROVED** [review]
🚢 it!

**Pull request merged**